src/vs/workbench/contrib/search/browser/searchModel.ts

```diff
@@ -2279,6 +2419,63 @@ export class RangeHighlightDecorations implements IDisposable {
 	});
 }
 
+interface INotebookDataEditInfo {
+	notebookData: NotebookData;
+	mTime: number;
+}
+class NotebookDataCache {
+	private _entries: ResourceMap<INotebookDataEditInfo>;
+	// private _serializer: INotebookSerializer | undefined;
+
+	constructor(
+		@IUriIdentityService private readonly uriIdentityService: IUriIdentityService,
+		@IFileService private readonly fileService: IFileService,
+		@INotebookService private readonly notebookService: INotebookService,
+		@IEditorResolverService private readonly editorResolverService: IEditorResolverService,
+	) {
+		this._entries = new ResourceMap<INotebookDataEditInfo>(uri => this.uriIdentityService.extUri.getComparisonKey(uri));
+	}
+
+	async getSerializer(notebookUri: URI): Promise<INotebookSerializer | undefined> {
+		const registeredEditorInfo = this.editorResolverService.getEditors(notebookUri);
+		const priorityEditorInfo = registeredEditorInfo.reduce((acc, val) =>
+			priorityToRank(acc.priority) > priorityToRank(val.priority) ? acc : val
+		);
+		const info = await this.notebookService.withNotebookDataProvider(priorityEditorInfo.id);
+		if (!(info instanceof SimpleNotebookProviderInfo)) {
+			return undefined;
+		}
+		return info.serializer;
+	}
+
+	async getNotebookData(notebookUri: URI): Promise<NotebookData> {
+		const mTime = (await this.fileService.stat(notebookUri)).mtime;
```

Comment:
There are a lot of back and forth calls to the EH for this code to run and I'm wondering whether more of it could actually live in the extension host process. That could hugely cut down on your latency. However it might be hard to interact with notebook serializers from the EH (even though that's where they live) and the code that does things like checking whether a notebook is already open and using its model of course has to stay in the renderer process.